File: dailyQuestion/2020-09/9-18/golang/solution_dfs.go
package main

import (
	"fmt"
)

func main() {
	result := countNumbersWithUniqueDigits(2)
	fmt.Println(result)
}

func countNumbersWithUniqueDigits(n int) int {
	if n == 0 {
		return 1
	}
	var dfs func(idx int) int
	used := make([]bool, 10)
	dfs = func(idx int) int {
		result := 0
		if idx == n {
			return 0
		}
		for num := 0; num < 10; num++ {
			// 剪枝：多位数时，第一个数字不能为0
			if n >= 2 && idx == 1 && num == 0 {
				continue
			}
			// 剪枝：不能使用用过的数字
			if used[num] {
				continue
			}
			used[num] = true
			result += dfs(idx+1) + 1
			used[num] = false
		}
		return result
	}
	return dfs(0)
}

File: dailyQuestion/2020-09/9-18/python/solution_dfs.py
def countNumbersWithUniqueDigits(n: int) -> int:
    if n == 0 :
        return 1
    used = [False] * 10
    def dfs(idx: int) -> int:
        if idx == n :
            return 0
        result = 0
        for num in range(10) :
            if n >= 2 and idx == 1 and num == 0 :
                continue
            if used[num] :
                continue
            used[num] = True
            result += dfs(idx + 1) + 1
            used[num] = False
        return result
    return dfs(0)

if __name__ == "__main__" :
    result = countNumbersWithUniqueDigits(2)
    print(result)